query-sql/src/lib.rs

```diff
@@ -290,19 +322,15 @@ impl QueryFragment for SelectQuery {
     fn push_sql(&self, out: &mut QueryBuilder) -> BuildQueryResult {
         out.push_sql("SELECT ");
         self.projection.push_sql(out)?;
-
-        out.push_sql(" FROM ");
         self.from.push_sql(out)?;
 
-        if self.constraints.is_empty() {
-            return Ok(());
+        if !self.constraints.is_empty() {
```

Comment:
I see what you mean. As I noted above: it's possible to have constraints and/or a projection when `from` is `None`. Indeed, in Datalog `ground` is `SELECT 123 AS x`, and `not` is effectively `SELECT 1 WHERE $condition`.